### PR TITLE
Refactorings in preparation for incremental loading

### PR DESCRIPTION
**Pull request by wemrysi**
This is a collection of changes needed in preparation for incremental loading:

* Unifying `DestinationColumn` and `TableColumn` into `quasar.api.Column`
* Removing some unused datasource implementations
* Reorganizing `connector` by adding `datasource`, `destination` and `render` subpackages
* Relocating `Destination`, `ResultRender` and associated types to `connector` module
* Refactoring `Datasource` and `Destination` to be extensible for incremental loading and hopefully reduce churn for future additions

### CODE REVIEW

**Comment by wemrysi** [comment]
I'd recommend reviewing per-commit, I chose to group these into a single PR as they're all contributing to the incremental loading effort and to reduce the number of breaking changes that need to be shepherded through the stack.

File: api/src/main/scala/quasar/api/destination/Destinations.scala [hunk 1]
```diff
@@ -44,9 +44,6 @@ trait Destinations[F[_], G[_], I, C] {
     */
   def destinationRef(destinationId: I): F[ExistentialError[I] \/ DestinationRef[C]]
 
-  /** Returns the specified destination. Initializes it if necessary */
-  def destinationOf(destinationId: I): F[DestinationError[I, C] \/ Destination[F]]
```
Comment (alissapajer) [inline api/src/main/scala/quasar/api/destination/Destinations.scala]: Why was this deleted?

**Comment by alissapajer** [comment]
@wemrysi you can merge this so that it lands when you're ready to do the shepherding.

File: api/src/main/scala/quasar/api/destination/Destinations.scala (hunk 1 above)
Comment (wemrysi) [reply api/src/main/scala/quasar/api/destination/Destinations.scala]: `Destination` isn't part of the client api (their functionality is exposed via `ResultPush`), they were never really safe to expose since their lifecycle is managed elsewhere, so a caller of this function could obtain a reference to a destination which may later be finalized and end up in an undefined state.

Making this explicit also allowed us to move all the destination-related types into the `connector` module and out of `api`, which was one of the main goals of this refactoring.

File: api/src/main/scala/quasar/api/Column.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+object Column {
+  implicit val traverseColumn: Traverse[Column] =
+    new Traverse[Column] {
+      def foldLeft[A, B](fa: Column[A], b: B)(f: (B, A) => B): B =
+        f(b, fa.tpe)
+
+      def foldRight[A, B](fa: Column[A], lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B] =
+        f(fa.tpe, lb)
+
+      override def map[A, B](fa: Column[A])(f: A => B): Column[B] =
```
Comment (alissapajer) [inline api/src/main/scala/quasar/api/Column.scala]: @wemrysi why the override of `map`? Given [this override](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/Traverse.scala#L95-L96) plus this definition of [`Id#map`](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/package.scala#L48), what's the gain?

Comment (wemrysi) [reply api/src/main/scala/quasar/api/Column.scala]: Not much gain in this case, I didn't even notice, tbh. I copied the impl from the deleted `DestinationColumn`.
